src/edu/csus/ecs/pc2/clics/API202306/ProblemService.java

```diff
@@ -0,0 +1,143 @@
+        if(contestId.equals(model.getContestIdentifier()) == false) {
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+        // get the problems from the contest
+        Problem[] problems = model.getProblems();
+        int ord = 1;
+        ArrayList<CLICSProblem> problist = new ArrayList<CLICSProblem>();
+
+        // public only gets the problems when the contest starts
+        if (sc.isUserInRole(WebServer.WEBAPI_ROLE_ADMIN) || sc.isUserInRole(WebServer.WEBAPI_ROLE_JUDGE) || model.getContestTime().isContestStarted()) {
+            for (Problem problem: problems) {
+                if (problem.isActive()) {
+                    problist.add(new CLICSProblem(model, problem, ord));
+                    ord++;
+                }
+            }
+        }
+        try {
+            ObjectMapper mapper = JSONUtilities.getObjectMapper();
+            String json = mapper.writeValueAsString(problist);
+            return Response.ok(json, MediaType.APPLICATION_JSON).build();
+        } catch (Exception e) {
+            return Response.status(Status.INTERNAL_SERVER_ERROR).entity("Error creating JSON for problems " + e.getMessage()).build();
+        }
+    }
+
+    /**
+     * Return the reponse to a request for a single problem's information for the specified contest and problem id.
+     *
+     * @param sc user info
+     * @param contestId the contest
+     * @param problemId the problem id desired
+     * @return response
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("{problemId}/")
+    public Response getProblem(@Context SecurityContext sc, @PathParam("contestId") String contestId, @PathParam("problemId") String problemId) {
+        // public only gets the problems when the contest starts
+        if (sc.isUserInRole(WebServer.WEBAPI_ROLE_ADMIN) || sc.isUserInRole(WebServer.WEBAPI_ROLE_JUDGE) || model.getContestTime().isContestStarted()) {
+            // get the problems from the contest
+            Problem[] problems = model.getProblems();
+            int ord = 1;
+            for (Problem problem: problems) {
+                // match by ID
+                if (problem.isActive()) {
+                    if(IJSONTool.getProblemId(problem).equals(problemId)) {
+                        try {
+                            return Response.ok(JSONUtilities.getObjectMapper().writeValueAsString(new CLICSProblem(model, problem, ord)), MediaType.APPLICATION_JSON).build();
+                        } catch(Exception e) {
+                            return Response.status(Status.INTERNAL_SERVER_ERROR).entity("Error creating JSON for problem " + problemId + " in contest " + contestId + ": " + e.getMessage()).build();
+                        }
+                    }
+                    ord++;
+                }
+            }
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+        return Response.status(Response.Status.FORBIDDEN).build();
```

Comment:
THIS is what I expected to see in method `getProblems()` when the invoker isn't allowed to see the problem(s).

Reply:
This is different.  Someone is requesting a specific(!) problem, not the list. It means that at some point, they got the list (since they know the problem ID).  If they are allowed to see the problem, it will be returned, otherwise, it could be someone attempting to guess a problem ID, so we return forbidden.